File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    
    def isOpening(self, c):
        if c in "({[":
            return True
        return False
    
    def isMatch(self, a, b):
        d = {
            "(": ")",
            "{": "}",
            "[": "]"
        }
        
        if d[a] == b:
            return True
        return False
        
    def isValid(self, s: str) -> bool:
        
        if len(s)%2 == 1:
            return False
        
        stack = []
        
        for each in s:
            if len(stack) != 0 and self.isOpening(each):
                stack.append(each)
                continue
            if self.isOpening(each):
                stack.append(each)
            elif len(stack) > 0 and self.isMatch(stack[len(stack)-1], each):
                stack.pop()
            else:
                return False
        if stack:
            return False
        return True
            

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    private boolean isOpening(char c) {
        return c == '(' || c ==  '{' || c ==  '[';
    }
    
    private boolean isMatching(char open, char close) {
        Map<Character, Character> map = new HashMap<Character, Character>();
        
        map.put('(', ')');
        map.put('{', '}');
        map.put('[', ']');
        
        if (map.get(open) == close) {
            return true;
        } 
        return false;
    }
    public boolean isValid(String s) {
        
        if (s.length() % 2 == 1) {
            return false;
        }
        
        Stack<Character> stack = new Stack<>();
        
        for (int i=0; i<s.length(); i++) {
            if (!stack.isEmpty() && isOpening(s.charAt(i))) {
                stack.push(s.charAt(i));
                continue;
            }
            if (isOpening(s.charAt(i))) {
                stack.push(s.charAt(i));
            }
            else if (!stack.isEmpty() && isMatching(stack.peek(), s.charAt(i))) {
                stack.pop();
            } else {
                return false;
            }
        }
    
        if (!stack.isEmpty()) {
            return false;
        }
        return true;
    }
}